sklearn/preprocessing/tests/test_encoders.py

```diff
@@ -637,3 +652,314 @@ def test_categories(density, drop):
 @pytest.mark.parametrize('Encoder', [OneHotEncoder, OrdinalEncoder])
 def test_encoders_has_categorical_tags(Encoder):
     assert 'categorical' in Encoder()._get_tags()['X_types']
+
+
+@pytest.mark.parametrize("kwargs", [
+    {'max_levels': 2},
+    {'min_frequency': 11},
+    {'min_frequency': 0.29},
+    {'max_levels': 2, 'min_frequency': 6},
+    {'max_levels': 4, 'min_frequency': 12},
+])
+@pytest.mark.parametrize("categories",
+                         ["auto", [['a', 'b', 'c', 'd']]])
+def test_ohe_infrequent_two_levels(kwargs, categories):
+
+    X_train = np.array([['a'] * 5 + ['b'] * 20 + ['c'] * 10 + ['d'] * 3]).T
+    ohe = OneHotEncoder(categories=categories,
+                        handle_unknown='auto', sparse=False,
+                        **kwargs).fit(X_train)
+    assert_array_equal(ohe.infrequent_indices_, [[0, 2, 3]])
+
+    X_test = [['b'], ['a'], ['c'], ['d'], ['e']]
+    expected = np.array([
+        [1, 0],
+        [0, 1],
+        [0, 1],
+        [0, 1],
+        [0, 1]])
+
+    X_trans = ohe.transform(X_test)
+    assert_allclose(expected, X_trans)
+
+    expected_inv = [['b'], ['c'], ['c'], ['c'], ['c']]
+    X_inv = ohe.inverse_transform(X_trans)
+    assert_array_equal(expected_inv, X_inv)
+
+    # The most frequent infrequent category becomes the feature name
+    feature_names = ohe.get_feature_names()
+    assert_array_equal(['x0_b', 'x0_c'], feature_names)
+
+
+@pytest.mark.parametrize("kwargs", [
+    {'max_levels': 3},
+    {'min_frequency': 6},
+    {'min_frequency': 9},
+    {'min_frequency': 0.24},
+    {'min_frequency': 0.16},
+    {'max_levels': 3, 'min_frequency': 8},
+    {'max_levels': 4, 'min_frequency': 6},
+])
+def test_ohe_infrequent_three_levels(kwargs):
+
+    X_train = np.array([['a'] * 5 + ['b'] * 20 + ['c'] * 10 + ['d'] * 3]).T
+    ohe = OneHotEncoder(handle_unknown='auto', sparse=False,
+                        **kwargs).fit(X_train)
+    assert_array_equal(ohe.infrequent_indices_, [[0, 3]])
+
+    X_test = [['b'], ['a'], ['c'], ['d'], ['e']]
+    expected = np.array([
+        [1, 0, 0],
+        [0, 0, 1],
+        [0, 1, 0],
+        [0, 0, 1],
+        [0, 0, 1]])
+
+    X_trans = ohe.transform(X_test)
+    assert_allclose(expected, X_trans)
+
+    expected_inv = [['b'], ['a'], ['c'], ['a'], ['a']]
+    X_inv = ohe.inverse_transform(X_trans)
+    assert_array_equal(expected_inv, X_inv)
+
+    # The most frequent infrequent category becomes the feature name
+    feature_names = ohe.get_feature_names()
+    assert_array_equal(['x0_b', 'x0_c', 'x0_a'], feature_names)
+
+
+def test_ohe_infrequent_two_levels_user_cats():
```

Comment:
I would prefer the relationship between the behaviour for user-specified cats and auto cats to be a bit clearer here. At the moment this tests the behaviour independently, rather than asserting that some relationship holds between the two behaviours. Is there a nice way to test that relationship instead?

Reply:
All the other test cases uses auto and checks that the `infrequent_indices_` respects the lexicon ordering from auto.

The biggest different is how user provided cats may not have the same order as 'auto'. What relationship do you see between them?